### PR TITLE
Guided decoding

### PR DESCRIPTION
**Pull request by yunfeng-scale**
# Pull Request Summary

Support guided_json, guided_regex, guided_choice from vLLM

## Test Plan and Usage Guide

tested within vllm entrypoint and from model engine
```
curl localhost:5005/predict -d '{"prompt":"hello", "guided_regex":"a beautiful world .*"}'
{"text": "a beautiful world is fuxix", "count_prompt_tokens": 2, "count_output_tokens": 16, "log_probs": null, "tokens": ["a", " ", "b", "ea", "u", "t", "iful", " ", "wo", "r", "l", "d", " is", " f", "ux", "ix"]}
```

### CODE REVIEW

File: model-engine/model_engine_server/domain/use_cases/llm_model_endpoint_use_cases.py [hunk 1]
```diff
@@ -1365,6 +1365,11 @@ def validate_and_update_completion_params(
                 "include_stop_str_in_output is only supported in vllm."
             )
 
+    if (
+        request.guided_choice or request.guided_regex or request.guided_json
+    ) and not inference_framework == LLMInferenceFramework.VLLM:
```
Comment (saiatmakuri) [inline model-engine/model_engine_server/domain/use_cases/llm_model_endpoint_use_cases.py:1370]: also check that only 1 of the following is provided?

File: model-engine/model_engine_server/inference/vllm/vllm_server.py [hunk 2]
```diff
@@ -38,7 +40,30 @@ async def generate(request: Request) -> Response:
     request_dict = await request.json()
     prompt = request_dict.pop("prompt")
     stream = request_dict.pop("stream", False)
+    guided_json = request_dict.pop("guided_json", None)
+    guided_regex = request_dict.pop("guided_regex", None)
+    guided_choice = request_dict.pop("guided_choice", None)
     sampling_params = SamplingParams(**request_dict)
+
+    # Dummy request to get guided decode logit processor
+    partial_openai_request = OpenAICompletionRequest.model_validate(
```
Comment (saiatmakuri) [inline model-engine/model_engine_server/inference/vllm/vllm_server.py:49]: a validation error here would be thrown as a 500 right? lets make sure its a 400 instead?

**Review by saiatmakuri — APPROVED** [review]

File: model-engine/model_engine_server/domain/use_cases/llm_model_endpoint_use_cases.py (hunk 1 above)
Comment (yunfeng-scale) [reply model-engine/model_engine_server/domain/use_cases/llm_model_endpoint_use_cases.py:1370]: sure

File: model-engine/model_engine_server/inference/vllm/vllm_server.py (hunk 2 above)
Comment (yunfeng-scale) [reply model-engine/model_engine_server/inference/vllm/vllm_server.py:49]: sure. although validation failure should happen upstream at gateway

**Pull request merged**